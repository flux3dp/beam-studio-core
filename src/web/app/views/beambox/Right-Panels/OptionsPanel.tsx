import React, { useMemo } from 'react';

import ColorPanel from 'app/views/beambox/Right-Panels/ColorPanel';
import ImageOptions from 'app/views/beambox/Right-Panels/Options-Blocks/Image-Options';
import InFillBlock from 'app/views/beambox/Right-Panels/Options-Blocks/InFillBlock';
import MultiColorOptions from 'app/views/beambox/Right-Panels/Options-Blocks/MultiColorOptions';
import ObjectPanelItem from 'app/views/beambox/Right-Panels/ObjectPanelItem';
import PolygonOptions from 'app/views/beambox/Right-Panels/Options-Blocks/PolygonOptions';
import RectOptions from 'app/views/beambox/Right-Panels/Options-Blocks/RectOptions';
import TextOptions from 'app/views/beambox/Right-Panels/Options-Blocks/TextOptions';
import { getObjectLayer } from 'helpers/layer/layer-helper';
import { useIsMobile } from 'helpers/system-helper';

import styles from './OptionsPanel.module.scss';

interface Props {
  elem: Element;
  rx: number;
  polygonSides?: number;
  updateObjectPanel: () => void;
  updateDimensionValues: (val: { [key: string]: any }) => void;
}

function OptionsPanel({
  elem,
  rx,
  polygonSides,
  updateObjectPanel,
  updateDimensionValues,
}: Props): JSX.Element {
  const isMobile = useIsMobile();
  let contents: JSX.Element | JSX.Element[];
  const showColorPanel = useMemo(() => {
    if (!elem) return false;
    if (!['rect', 'ellipse', 'path', 'text', 'polygon', 'g', 'use'].includes(elem?.tagName.toLowerCase())) return false;
    return getObjectLayer(elem as SVGElement)?.elem?.getAttribute('data-fullcolor') === '1';
  }, [elem]);
  if (elem) {
    const tagName = elem.tagName.toLowerCase();
    if (tagName === 'rect') {
      contents = [
        <RectOptions
          key="rect"
          elem={elem}
          rx={rx}
          updateDimensionValues={updateDimensionValues}
        />,
        showColorPanel ? <ColorPanel elem={elem} /> : <InFillBlock key="fill" elem={elem} />,
      ];
    } else if (tagName === 'polygon') {
      contents = [
        <PolygonOptions key="polygon" elem={elem} polygonSides={polygonSides} />,
        showColorPanel ? <ColorPanel elem={elem} /> : <InFillBlock key="fill" elem={elem} />,
      ];
    } else if (tagName === 'text') {
      contents = [
        <TextOptions
          key="text"
          elem={elem}
          textElement={elem as SVGTextElement}
          updateObjectPanel={updateObjectPanel}
          updateDimensionValues={updateDimensionValues}
        />,
        showColorPanel ? <ColorPanel elem={elem} /> : <InFillBlock key="fill" elem={elem} />,
      ];
    } else if (tagName === 'image' || tagName === 'img') {
      if (elem.getAttribute('data-fullcolor') === '1') contents = null;
      else contents = <ImageOptions elem={elem} updateObjectPanel={updateObjectPanel} />;
    } else if (tagName === 'g' && elem.getAttribute('data-textpath-g')) {
      const textElem = elem.querySelector('text');
      contents = (
        <TextOptions
          isTextPath
          elem={elem}
          textElement={textElem}
          updateObjectPanel={updateObjectPanel}
          updateDimensionValues={updateDimensionValues}
        />
      );
    } else if (tagName === 'g') {
      contents = showColorPanel ? <MultiColorOptions elem={elem} /> : <InFillBlock elem={elem} />;
    } else if (tagName === 'use') {
      contents = showColorPanel ? <MultiColorOptions elem={elem} /> : null;
    } else {
      contents = showColorPanel ? <ColorPanel elem={elem} /> : <InFillBlock elem={elem} />;
    }
  }

  return isMobile ? (
    <div className={styles.container}>
      <ObjectPanelItem.Divider />
      {contents}
    </div>
  ) : (
    <>
      {(contents) && (
        <div className={styles.panel}>
          <div className={styles.title}>OPTIONS</div>
          {contents}
      </div>
      )}
    </>
  );
}

export default OptionsPanel;
