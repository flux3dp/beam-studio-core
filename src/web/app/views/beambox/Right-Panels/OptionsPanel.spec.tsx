/* eslint-disable import/first */
import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('app/views/beambox/Right-Panels/Options-Blocks/Image-Options', () => function ImageOptions() {
  return (
    <div>
      This is dummy ImageOptions
    </div>
  );
});

jest.mock('app/views/beambox/Right-Panels/Options-Blocks/InFillBlock', () => function DummyInFillBlock() {
  return (
    <div>
      This is dummy InFillBlock
    </div>
  );
});

jest.mock('app/views/beambox/Right-Panels/Options-Blocks/RectOptions', () => function RectOptions() {
  return (
    <div>
      This is dummy RectOptions
    </div>
  );
});

jest.mock('app/views/beambox/Right-Panels/Options-Blocks/TextOptions', () => function TextOptions() {
  return (
    <div>
      This is dummy TextOptions
    </div>
  );
});

jest.mock('app/views/beambox/Right-Panels/Options-Blocks/PolygonOptions', () => function PolygonOptions() {
  return (
    <div>
      This is dummy PolygonOptions
    </div>
  );
});

import OptionsPanel from './OptionsPanel';

describe('should render correctly', () => {
  test('rect', () => {
    const updateDimensionValues = jest.fn();
    document.body.innerHTML = '<rect id="rect" />';
    const wrapper = shallow(<OptionsPanel
      elem={document.getElementById('rect')}
      rx={null}
      updateObjectPanel={jest.fn()}
      updateDimensionValues={updateDimensionValues}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('RectOptions').props().updateDimensionValues();
    expect(updateDimensionValues).toHaveBeenCalledTimes(1);
  });

  test('text', () => {
    const updateObjectPanel = jest.fn();
    const updateDimensionValues = jest.fn();
    document.body.innerHTML = '<text id="text" />';
    const wrapper = shallow(<OptionsPanel
      elem={document.getElementById('text')}
      rx={null}
      updateObjectPanel={updateObjectPanel}
      updateDimensionValues={updateDimensionValues}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('TextOptions').props().updateDimensionValues();
    expect(updateDimensionValues).toHaveBeenCalledTimes(1);

    wrapper.find('TextOptions').props().updateObjectPanel();
    expect(updateObjectPanel).toHaveBeenCalledTimes(1);
  });

  test('image', () => {
    const updateObjectPanel = jest.fn();
    document.body.innerHTML = '<image id="image" />';
    const wrapper = shallow(<OptionsPanel
      elem={document.getElementById('image')}
      rx={null}
      updateObjectPanel={updateObjectPanel}
      updateDimensionValues={jest.fn()}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('ImageOptions').props().updateObjectPanel();
    expect(updateObjectPanel).toHaveBeenCalledTimes(1);
  });

  describe('polygon', () => {
    test('desktop version', () => {
      document.body.innerHTML = '<polygon id="polygon" />';
      const wrapper = shallow(<OptionsPanel
        elem={document.getElementById('polygon')}
        rx={null}
        polygonSides={8}
        updateObjectPanel={jest.fn()}
        updateDimensionValues={jest.fn()}
      />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('web version', () => {
      window.FLUX.version = 'web';
      document.body.innerHTML = '<polygon id="polygon" />';
      const wrapper = shallow(<OptionsPanel
        elem={document.getElementById('polygon')}
        rx={null}
        polygonSides={8}
        updateObjectPanel={jest.fn()}
        updateDimensionValues={jest.fn()}
      />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  test('others', () => {
    document.body.innerHTML = '<xxx id="xxx" />';
    const wrapper = shallow(<OptionsPanel
      elem={document.getElementById('xxx')}
      rx={null}
      updateObjectPanel={jest.fn()}
      updateDimensionValues={jest.fn()}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('no element', () => {
    const wrapper = shallow(<OptionsPanel
      elem={null}
      rx={null}
      updateObjectPanel={jest.fn()}
      updateDimensionValues={jest.fn()}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
