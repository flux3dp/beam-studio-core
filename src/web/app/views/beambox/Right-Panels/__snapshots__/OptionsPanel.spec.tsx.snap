// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly image 1`] = `
<div
  className="panel"
>
  <div
    className="title"
  >
    OPTIONS
  </div>
  <ImageOptions
    elem={
      <img
        id="image"
      />
    }
    updateObjectPanel={[MockFunction]}
  />
</div>
`;

exports[`should render correctly in mobile image 1`] = `
<div
  className="container"
>
  <ObjectPanelDivider />
  <ImageOptions
    elem={
      <img
        id="image"
      />
    }
    updateObjectPanel={[MockFunction]}
  />
</div>
`;

exports[`should render correctly in mobile no element 1`] = `""`;

exports[`should render correctly in mobile others 1`] = `
<div
  className="container"
>
  <ObjectPanelDivider />
  <DummyInFillBlock
    elem={
      <xxx
        id="xxx"
      />
    }
  />
</div>
`;

exports[`should render correctly in mobile polygon desktop version 1`] = `
<div
  className="container"
>
  <ObjectPanelDivider />
  <PolygonOptions
    elem={
      <polygon
        id="polygon"
      />
    }
    polygonSides={8}
  />
</div>
`;

exports[`should render correctly in mobile polygon web version 1`] = `
<div
  className="container"
>
  <ObjectPanelDivider />
  <PolygonOptions
    elem={
      <polygon
        id="polygon"
      />
    }
    polygonSides={8}
  />
</div>
`;

exports[`should render correctly in mobile rect 1`] = `
<div
  className="container"
>
  <ObjectPanelDivider />
  <RectOptions
    elem={
      <rect
        id="rect"
      />
    }
    rx={null}
    updateDimensionValues={[MockFunction]}
  />
</div>
`;

exports[`should render correctly in mobile text 1`] = `
<div
  className="container"
>
  <ObjectPanelDivider />
  <TextOptions
    elem={
      <text
        id="text"
      />
    }
    textElement={
      <text
        id="text"
      />
    }
    updateDimensionValues={[MockFunction]}
    updateObjectPanel={[MockFunction]}
  />
</div>
`;

exports[`should render correctly no element 1`] = `""`;

exports[`should render correctly others 1`] = `
<div
  className="panel"
>
  <div
    className="title"
  >
    OPTIONS
  </div>
  <DummyInFillBlock
    elem={
      <xxx
        id="xxx"
      />
    }
  />
</div>
`;

exports[`should render correctly polygon desktop version 1`] = `
<div
  className="panel"
>
  <div
    className="title"
  >
    OPTIONS
  </div>
  <PolygonOptions
    elem={
      <polygon
        id="polygon"
      />
    }
    polygonSides={8}
  />
</div>
`;

exports[`should render correctly polygon web version 1`] = `
<div
  className="panel"
>
  <div
    className="title"
  >
    OPTIONS
  </div>
  <PolygonOptions
    elem={
      <polygon
        id="polygon"
      />
    }
    polygonSides={8}
  />
</div>
`;

exports[`should render correctly rect 1`] = `
<div
  className="panel"
>
  <div
    className="title"
  >
    OPTIONS
  </div>
  <RectOptions
    elem={
      <rect
        id="rect"
      />
    }
    rx={null}
    updateDimensionValues={[MockFunction]}
  />
</div>
`;

exports[`should render correctly text 1`] = `
<div
  className="panel"
>
  <div
    className="title"
  >
    OPTIONS
  </div>
  <TextOptions
    elem={
      <text
        id="text"
      />
    }
    textElement={
      <text
        id="text"
      />
    }
    updateDimensionValues={[MockFunction]}
    updateObjectPanel={[MockFunction]}
  />
</div>
`;
